pkg/kubelet/certificate/certificate_manager.go

```diff
@@ -0,0 +1,371 @@
+	// TLS server will automatically call back here to get the correct
+	// certificate when establishing each new connection.
+	GetCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error)
+}
+
+type manager struct {
+	certSigningRequestClient clientcertificates.CertificateSigningRequestInterface
+	nodeName                 types.NodeName
+	certAccessLock           sync.Mutex
+	certDirectory            string
+	keyDirectory             string
+	cert                     *tls.Certificate
+	certRotationEnabled      bool
+}
+
+// NewManager returns a new certificate manager. A certificate manager is
+// responsible for being the authoritative source of certificates in the
+// Kubelet and handling updates due to rotation.
+func NewManager(
+	certSigningRequestClient clientcertificates.CertificateSigningRequestInterface,
+	nodeName types.NodeName,
+	certDirectory string,
+	keyDirectory string,
+	certFile string,
+	keyFile string,
+	certRotationEnabled bool) (Manager, error) {
+
+	var selectedCertFile, selectedKeyFile string
+	if fileExists(certFile) && fileExists(keyFile) {
+		selectedCertFile = certFile
+		selectedKeyFile = keyFile
+	} else if fileExists(filepath.Join(certDirectory, pairNamePrefix+certExtension)) && fileExists(filepath.Join(keyDirectory, pairNamePrefix+keyExtension)) {
+		selectedCertFile = filepath.Join(certDirectory, pairNamePrefix+certExtension)
+		selectedKeyFile = filepath.Join(keyDirectory, pairNamePrefix+keyExtension)
+	} else {
+		return nil, fmt.Errorf("no cert/key files read at (%q, %q) or in (%q, %q)",
+			certFile,
+			keyFile,
+			certDirectory,
+			keyDirectory)
+	}
+
+	glog.Infof("Loading cert/key data from (%q, %q)", selectedCertFile, selectedKeyFile)
+	cert, err := loadX509KeyPair(selectedCertFile, selectedKeyFile)
+	if err != nil {
+		return nil, err
+	}
+
+	m := manager{
+		certSigningRequestClient: certSigningRequestClient,
+		nodeName:                 nodeName,
+		certDirectory:            filepath.Dir(selectedCertFile),
+		keyDirectory:             filepath.Dir(selectedKeyFile),
+		certRotationEnabled:      certRotationEnabled,
+		cert:                     &cert,
+	}
+
+	if m.certRotationEnabled {
+		if err := m.cleanupPairs(); err == nil {
+			return &m, nil
+		} else {
+			return nil, err
+		}
+	}
+	return &m, nil
+}
+
+// GetCertificate returns the certificate that should be used TLS connections.
+// The value returned by this function will change over time as the certificate
+// is rotated. If a reference to this method is passed directly into the TLS
+// options for a connection, certificate rotation will be handled correctly by
+// the underlying go libraries.
+//
+//    tlsOptions := &server.TLSOptions{
+//        ...
+//        GetCertificate: certificateManager.GetCertificate
+//        ...
+//    }
+//
+func (m *manager) GetCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	m.certAccessLock.Lock()
+	defer m.certAccessLock.Unlock()
+	return m.cert, nil
```

Comment:
For such a rarely written value, maybe a lockless atomic.Value Load() would be better